File: Python/Visitor/Gemini 1.5 Flash/round3/code.py
from abc import ABC, abstractmethod
from typing import List, Optional

class Account:
    def __init__(self, account_number: str, pin: str, balance: float = 0):
        self.account_number = account_number
        self.pin = pin
        self.balance = balance

    def get_balance(self) -> float:
        return self.balance

    def withdraw(self, amount: float) -> bool:
        if self.balance >= amount:
            self.balance -= amount
            return True
        return False

    def deposit(self, amount: float) -> None:
        self.balance += amount

    def change_pin(self, new_pin: str) -> None:
        self.pin = new_pin

class ATM:
    def __init__(self):
        self.accounts = {}  # Example: {"account_number": Account}
        self.max_pin_attempts = 3
        self.transaction_history = []

    def add_account(self, account: Account) -> None:
        self.accounts[account.account_number] = account

    def authenticate(self, account_number: str, pin: str) -> Optional[Account]:
        account = self.accounts.get(account_number)
        if account and account.pin == pin:
            return account
        return None

    def process_transaction(self, account: Account, visitor: "ATMVisitor") -> None:
        visitor.visit(account, self)

    def get_transaction_history(self) -> List[str]:
        return self.transaction_history

class ATMVisitor(ABC):
    @abstractmethod
    def visit(self, account: Account, atm: ATM) -> None:
        pass

class BalanceDisplayVisitor(ATMVisitor):
    def visit(self, account: Account, atm: ATM) -> None:
        balance = account.get_balance()
        atm.transaction_history.append(f"Balance: {balance}")
        print(f"Your current balance is: {balance}")

class CashWithdrawalVisitor(ATMVisitor):
    def __init__(self, amount: float):
        self.amount = amount

    def visit(self, account: Account, atm: ATM) -> None:
        if account.withdraw(self.amount):
            atm.transaction_history.append(f"Withdrawal: {self.amount}")
            print(f"You have withdrawn {self.amount}")
        else:
            atm.transaction_history.append(f"Insufficient funds for withdrawal: {self.amount}")
            print("Insufficient funds.")

class DepositVisitor(ATMVisitor):
    def __init__(self, amount: float):
        self.amount = amount

    def visit(self, account: Account, atm: ATM) -> None:
        account.deposit(self.amount)
        atm.transaction_history.append(f"Deposit: {self.amount}")
        print(f"You have deposited {self.amount}")

class PINChangeVisitor(ATMVisitor):
    def __init__(self, new_pin: str):
        self.new_pin = new_pin

    def visit(self, account: Account, atm: ATM) -> None:
        # Simulate PIN confirmation (you would typically ask for new pin twice)
        if True:  # Assuming confirmation is successful
            account.change_pin(self.new_pin)
            atm.transaction_history.append(f"PIN changed successfully.")
            print("PIN changed successfully.")
        else:
            atm.transaction_history.append(f"PIN change failed.")
            print("PIN change failed. Please try again.")

class ReceiptPrintingVisitor(ATMVisitor):
    def visit(self, account: Account, atm: ATM) -> None:
        print("Printing receipt...")
        print("-" * 20)
        for transaction in atm.get_transaction_history():
            print(transaction)
        print("-" * 20)
        print("Thank you!")

def main():
    atm = ATM()
    atm.add_account(Account("1234567890", "1234", 1000))

    while True:
        account_number = input("Enter your account number: ")
        pin = input("Enter your PIN: ")
        account = atm.authenticate(account_number, pin)

        if account:
            print("Authentication successful!")
            while True:
                print("\nChoose an option:")
                print("1. View balance")
                print("2. Withdraw cash")
                print("3. Deposit cash")
                print("4. Change PIN")
                print("5. Print receipt")
                print("6. Exit")

                choice = input("Enter your choice: ")

                if choice == '1':
                    atm.process_transaction(account, BalanceDisplayVisitor())
                elif choice == '2':
                    amount = float(input("Enter amount to withdraw: "))
                    atm.process_transaction(account, CashWithdrawalVisitor(amount))
                elif choice == '3':
                    amount = float(input("Enter amount to deposit: "))
                    atm.process_transaction(account, DepositVisitor(amount))
                elif choice == '4':
                    new_pin = input("Enter new PIN: ")
                    atm.process_transaction(account, PINChangeVisitor(new_pin))
                elif choice == '5':
                    atm.process_transaction(account, ReceiptPrintingVisitor())
                elif choice == '6':
                    print("Exiting...")
                    break
                else:
                    print("Invalid choice.")
        else:
            print("Authentication failed.")

if __name__ == "__main__":
    main()

File: Java/Visitor/Gemini 1.5 Flash/round3/code.java
import java.util.HashMap;
import java.util.Map;

// Account class
class Account {
    private int accountNumber;
    private int pin;
    private double balance;
    private int pinAttempts;

    public Account(int accountNumber, int pin, double balance) {
        this.accountNumber = accountNumber;
        this.pin = pin;
        this.balance = balance;
        this.pinAttempts = 0;
    }

    public int getAccountNumber() {
        return accountNumber;
    }

    public int getPin() {
        return pin;
    }

    public double getBalance() {
        return balance;
    }

    public void setBalance(double balance) {
        this.balance = balance;
    }

    public int getPinAttempts() {
        return pinAttempts;
    }

    public void incrementPinAttempts() {
        this.pinAttempts++;
    }

    public void resetPinAttempts() {
        this.pinAttempts = 0;
    }

    public void changePin(int newPin) {
        this.pin = newPin;
    }
}

// ATM interface
interface ATM {
    void authenticate(Account account, int pin);
    void displayBalance(Account account);
    void withdraw(Account account, double amount);
    void deposit(Account account, double amount);
    void changePin(Account account, int newPin);
    void printReceipt(Account account, String transactionType, double amount);
}

// Concrete ATM implementation
class ConcreteATM implements ATM {
    private Map<Integer, Account> accounts;

    public ConcreteATM() {
        accounts = new HashMap<>();
        // Initialize accounts (replace with your data source)
        accounts.put(1234567890, new Account(1234567890, 1111, 1000.0));
        accounts.put(9876543210, new Account(9876543210, 2222, 500.0));
    }

    @Override
    public void authenticate(Account account, int pin) {
        if (account.getPin() == pin) {
            account.resetPinAttempts();
            System.out.println("Authentication successful.");
        } else {
            account.incrementPinAttempts();
            System.out.println("Incorrect PIN. Attempts remaining: " + (3 - account.getPinAttempts()));
            if (account.getPinAttempts() >= 3) {
                System.out.println("Card retained.");
            }
        }
    }

    @Override
    public void displayBalance(Account account) {
        System.out.println("Your current balance: $" + account.getBalance());
    }

    @Override
    public void withdraw(Account account, double amount) {
        if (account.getBalance() >= amount) {
            account.setBalance(account.getBalance() - amount);
            System.out.println("Cash withdrawn: $" + amount);
            printReceipt(account, "Withdrawal", amount);
        } else {
            System.out.println("Insufficient funds.");
        }
    }

    @Override
    public void deposit(Account account, double amount) {
        account.setBalance(account.getBalance() + amount);
        System.out.println("Cash deposited: $" + amount);
        printReceipt(account, "Deposit", amount);
    }

    @Override
    public void changePin(Account account, int newPin) {
        account.changePin(newPin);
        System.out.println("PIN changed successfully.");
    }

    @Override
    public void printReceipt(Account account, String transactionType, double amount) {
        // Replace with your receipt printing logic
        System.out.println("----------------------");
        System.out.println("Transaction: " + transactionType);
        System.out.println("Amount: $" + amount);
        System.out.println("Balance: $" + account.getBalance());
        System.out.println("----------------------");
    }
}

// Visitor interface for transaction operations
interface TransactionVisitor {
    void visit(Account account, ATM atm);
}

// Concrete visitors for different transactions
class BalanceVisitor implements TransactionVisitor {
    @Override
    public void visit(Account account, ATM atm) {
        atm.displayBalance(account);
    }
}

class WithdrawalVisitor implements TransactionVisitor {
    private double amount;

    public WithdrawalVisitor(double amount) {
        this.amount = amount;
    }

    @Override
    public void visit(Account account, ATM atm) {
        atm.withdraw(account, amount);
    }
}

class DepositVisitor implements TransactionVisitor {
    private double amount;

    public DepositVisitor(double amount) {
        this.amount = amount;
    }

    @Override
    public void visit(Account account, ATM atm) {
        atm.deposit(account, amount);
    }
}

class ChangePinVisitor implements TransactionVisitor {
    private int newPin;

    public ChangePinVisitor(int newPin) {
        this.newPin = newPin;
    }

    @Override
    public void visit(Account account, ATM atm) {
        atm.changePin(account, newPin);
    }
}

// ATM system
public class ATMSystem {
    private ConcreteATM atm;

    public ATMSystem() {
        atm = new ConcreteATM();
    }

    public void run() {
        Scanner scanner = new Scanner(System.in);
        Account currentAccount = null;

        while (true) {
            System.out.println("\nATM Menu:");
            System.out.println("1. Insert Card");
            System.out.println("2. Exit");

            int choice = scanner.nextInt();
            scanner.nextLine(); // Consume newline

            if (choice == 1) {
                System.out.print("Enter your account number: ");
                int accountNumber = scanner.nextInt();
                scanner.nextLine();

                if (atm.accounts.containsKey(accountNumber)) {
                    currentAccount = atm.accounts.get(accountNumber);

                    while (true) {
                        System.out.println("\nATM Options:");
                        System.out.println("1. Balance Inquiry");
                        System.out.println("2. Withdraw Cash");
                        System.out.println("3. Deposit Cash");
                        System.out.println("4. Change PIN");
                        System.out.println("5. Logout");

                        choice = scanner.nextInt();
                        scanner.nextLine();

                        switch (choice) {
                            case 1:
                                atm.authenticate(currentAccount, currentAccount.getPin());
                                if (currentAccount.getPinAttempts() < 3) {
                                    new BalanceVisitor().visit(currentAccount, atm);
                                }
                                break;
                            case 2:
                                atm.authenticate(currentAccount, currentAccount.getPin());
                                if (currentAccount.getPinAttempts() < 3) {
                                    System.out.print("Enter withdrawal amount: $");
                                    double amount = scanner.nextDouble();
                                    scanner.nextLine();
                                    new WithdrawalVisitor(amount).visit(currentAccount, atm);
                                }
                                break;
                            case 3:
                                atm.authenticate(currentAccount, currentAccount.getPin());
                                if (currentAccount.getPinAttempts() < 3) {
                                    System.out.print("Enter deposit amount: $");
                                    double amount = scanner.nextDouble();
                                    scanner.nextLine();
                                    new DepositVisitor(amount).visit(currentAccount, atm);
                                }
                                break;
                            case 4:
                                atm.authenticate(currentAccount, currentAccount.getPin());
                                if (currentAccount.getPinAttempts() < 3) {
                                    System.out.print("Enter new PIN (4 digits): ");
                                    int newPin = scanner.nextInt();
                                    scanner.nextLine();
                                    new ChangePinVisitor(newPin).visit(currentAccount, atm);
                                }
                                break;
                            case 5:
                                currentAccount = null;
                                System.out.println("Logged out.");
                                break;
                            default:
                                System.out.println("Invalid option.");
                        }
                        if (choice == 5) {
                            break;
                        }
                    }
                } else {
                    System.out.println("Invalid account number.");
                }
            } else if (choice == 2) {
                System.out.println("Exiting ATM system.");
                break;
            } else {
                System.out.println("Invalid option.");
            }
        }
    }

    public static void main(String[] args) {
        ATMSystem atmSystem = new ATMSystem();
        atmSystem.run();
    }
}